utils/checks.stanza

```diff
@@ -179,34 +194,33 @@ pcb-check no-connect-pins-not-connected (module:JITXObject) :
       fail-message =         "Pin %_ is marked no-connect and is incorrectly connected to other pins" % [ref(p)],
       locators =             [p])
 
-; A check used to verify the electrical properties of 
+; A check used to verify the electrical properties of
 ; digital i/o pins.
-pcb-check digital-io-levels (module:JITXObject) : 
-  val all-pins = all-pins(module)
-  val io-pins  = to-tuple $ filter(is-digital-io?, all-pins)
-  for p in io-pins do : 
-    val all-connected-pins = all-pins 
-      $> filter{{connected?([p, _0]) and _0 != p}, _}
-      $> to-tuple
-    val connected-io-pins = all-connected-pins
-      $> filter{is-digital-io?, _}
-      $> to-tuple
-    val connected-out-pins = all-connected-pins 
-      $> filter{has-property?{_.digital-output}, _}
-      $> to-tuple
+pcb-check digital-io-levels (module:JITXObject) :
+  val io-pins = filter(is-digital-io?, all-populated-pins(module))
+
+  ; FIXME! Could we split them first in connected groups and then run the loop only once per connected group?
+  ; Issue:: some of the logic is assymetrical and considers this p as special in the connected group considered
+  for p in io-pins do :
+    val all-connected-pins = populated-connected-pins(p)
```

Comment:
We can get all instances of a module and then build a cache of pin => dnp info. Then `populated-connected-pins` can filter on non-denp?

Reply:
We need to create a hashset of visited pin and vector of connected groups